PYIC-7746: Embedded metrics with powertools

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Embedded metrics with powertools

### Why did it change

Powertools have recently added support for sending embedded metrics. This uses that to send CRI redirect and return metrics, to allow us to track completion rates.

This will automatically add things like x-ray trace ids, and lambda request ids to the generated logs. We can also get cold start metrics for free.

<img width="1440" alt="Screenshot 2024-12-13 at 12 49 27" src="https://github.com/user-attachments/assets/49e125f9-5ebb-4017-a1e5-5536b9ec3f2f" />

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-7746](https://govukverify.atlassian.net/browse/PYIC-7746)


[PYIC-7746]: https://govukverify.atlassian.net/browse/PYIC-7746?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-7746: Embedded metrics with powertools

Powertools have recently added support for sending embedded metrics.
This uses that to send CRI redirect and return metrics, to allow us to
track completion rates.

Wynndow commented on deploy/template.yaml:
```diff
@@ -51,6 +52,7 @@ Globals:
         DT_OPEN_TELEMETRY_ENABLE_INTEGRATION: "true"
         OTEL_INSTRUMENTATION_AWS_SDK_EXPERIMENTAL_USE_PROPAGATOR_FOR_MESSAGING: true
         JAVA_TOOL_OPTIONS: -XX:+TieredCompilation -XX:TieredStopAtLevel=1
+        POWERTOOLS_METRICS_NAMESPACE: !Sub CoreBackEmbeddedMetrics-${Environment}
```
Adding the environment suffix here is useful for separating dev-env metrics.

Wynndow commented on lambdas/process-cri-callback/src/main/java/uk/gov/di/ipv/core/processcricallback/ProcessCriCallbackHandler.java:
```diff
@@ -143,26 +143,9 @@ public ProcessCriCallbackHandler() {
     public APIGatewayProxyResponseEvent handleRequest(
             APIGatewayProxyRequestEvent input, Context context) {
-        // temporary logging to check which tracing headers are being used by dynatrace
-        if ("build".equals(configService.getEnvironmentVariable(EnvironmentVariable.ENVIRONMENT))) {
```
Temporary logging from an old PR which should have been removed. Naughty.

Wynndow commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/item/IpvSessionItem.java:
```diff
@@ -42,7 +42,7 @@ public class IpvSessionItem implements PersistenceItem {
     private String emailAddress;
     private ReverificationStatus reverificationStatus;
-    private List<String> stateStack = new ArrayList<>();
+    @Builder.Default private List<String> stateStack = new ArrayList<>();
```
This is fixing a compilation warning - lombok didn't like the mix of having a builder and having the field already initialised.

Joe-Edwards-GDS replied:
(Fine - although we basically never actually use the builder)

Wynndow replied:
I think I introduced in pretty recently which is why there probably isn't much adoption of it yet - it was useful for tests

Joe-Edwards-GDS commented on lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/BuildCriOauthRequestHandler.java:
```diff
@@ -148,6 +150,7 @@ public BuildCriOauthRequestHandler(ConfigService configService) {
     @Override
     @Tracing
     @Logging(clearState = true)
+    @Metrics(captureColdStart = true)
```
Does this capture cold start times in a metric? (Should we think about adding this to all our lambdas?)

Wynndow replied:
Yeah, if we go down this route that's probably a good idea

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java:
```diff
@@ -0,0 +1,63 @@
+
+@ExcludeFromGeneratedCoverageReport
+public class EmbeddedMetricHelper {
+    private static final MetricsLogger METRICS_LOGGER = MetricsUtils.metricsLogger();
```
I assume it's fine to keep this as a static variable - i.e. it's safe to reload across snapstart?

We assign properties and dimensions directly to the logger, do they get reset between invocations?

Wynndow replied:
Yeah, the values get reset by the flush that happend at the end of every lambda. The metrics look good after running multiple journeys.

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java:
```diff
@@ -0,0 +1,63 @@
+                LOG_GOVUK_SIGNIN_JOURNEY_ID.getFieldName(),
+                govukSigninJourneyId == null
+                        ? GOVUK_SIGNIN_JOURNEY_ID_DEFAULT_VALUE
+                        : govukSigninJourneyId);
```
(Same comment as the other review about Log4J context)

Wynndow replied:
Yeah - I think it should be easily do-able. And I'm not sure that's a large risk of a key-clash. I'll give it a go.

Wynndow replied:
![Screenshot 2024-12-17 at 10 02 05](https://github.com/user-attachments/assets/359c7200-1bf6-4ee3-a259-449c6a0b048f)

Looks like it works pretty good.

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/EmbeddedMetricHelper.java:
```diff
@@ -0,0 +1,63 @@
+        createMetrics(Map.of(CRI, cri), Map.of(CRI_RETURN, 1.0), govukSigninJourneyId);
+    }
+
+    private static void createMetrics(
```
(Same comment as other review)

Joe-Edwards-GDS reviewed (commented):
Not much in it, but this seems slightly better and avoids needing to worry about flushing, what do you think?

Wynndow committed:
PYIC-7746: Include log4j context in metric

This includes all the values in log4j's MDC in the log message created
by the metrics logger. This means we get all the journey IDs that it
knows about which might be useful for advanced queries. It also makes
the signatures of the actual metric methods cleaner.

Wynndow committed:
PYIC-7746: Create cold start metric in all lambdas

The powertools metric lib will record a metric for us for a cold start.
This enables that on all our lambdas.

As the lambdas all now have the same configuration for the aspect
configuration, I've moved it up to the lambda sub-project and applied it
to all sub-projects.
